### PR TITLE
Add nodes to handle types

### PR DESCRIPTION
**Pull request by jdconrad**
This PR adds 3 nodes to handle types defined by a front-end creating a Painless AST. These types are decided with data immutability in mind - hence the reason for more than a single node.

DType - base class to allow for a getType call
DUnresolvedType - stores a String to be resolved to a type
DResolvedType - stores a Class<?> representative of a type

Initially this accomplishes consistency in how types are resolved in Painless instead of having loose types across multiple nodes. This PR replaces the type in SDeclaration as an example of usage, and follow up PRs will replace them in all nodes that have defined types.

Secondarily, this will be used allow future Painless improvements including making the AST more generic with the removal of custom try/catch handling as an example because now if we know a type internally we can just use a DResolvedType node instead of being forced to define a String type and resolve it multiple times.

Finally, this will be helpful if/when we decide to create a separate IR for valid semantic trees.

Edit: Adding @stu-elastic for visibility.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+package org.elasticsearch.painless.node;
+
+import org.elasticsearch.painless.Location;
+import org.elasticsearch.painless.lookup.PainlessLookup;
+import org.elasticsearch.painless.lookup.PainlessLookupUtility;
+
+import java.util.Objects;
+
+public class DResolvedType extends DType {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java]: Please add javadoc.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    protected final Class<?> type;
+    protected final boolean check;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java]: `check` means check against the whitelist, correct?  Please add a comment.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    public DResolvedType resolveType(PainlessLookup painlessLookup) {
+        if (check == true) {
+            if (painlessLookup.getClasses().contains(type) == false) {
+                throw location.createError(new IllegalArgumentException(
+                        "cannot resolve type [" + PainlessLookupUtility.typeToCanonicalTypeName(type) + "]"));
+            }
+
+            return new DResolvedType(location, type, false);
+        } else {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java]: Can you do `if (check == false) { return this; }` to reduce the indention level?

**Review by stu-elastic — APPROVED** [review]
Looks good.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SDeclaration.java [hunk 4]
```diff
@@ -37,13 +37,13 @@
  */
 public final class SDeclaration extends AStatement {
 
-    private final String type;
+    private DType type;
```
Comment (rjernst) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SDeclaration.java]: Can you please add a comment here explaining why this is mutable and what are plans are moving forward here?

**Review by rjernst — APPROVED** [review]
LGTM

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SDeclaration.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SDeclaration.java]: Changed this to final instead. Will add mutable DResolvedType nodes where necessary as the conversions take place. These should be removed with the creation of a second tree.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/DResolvedType.java]: Done. Also renamed to `checkInLookup`.

**Comment by jdconrad** [comment]
@rjernst @stu-elastic Thanks for the reviews. I believe I've addressed the feedback, and I will commit as soon as CI passes.

**Pull request merged**